test/robustness/within/quadtree_partition.hpp

```diff
@@ -0,0 +1,149 @@
+// Boost.Geometry (aka GGL, Generic Geometry Library)
+
+// Copyright (c) 2023 Barend Gehrels, Amsterdam, the Netherlands.
+// Use, modification and distribution is subject to the Boost Software License,
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_GEOMETRY_EXPERIMENTAL_QUADTREE_PARTITION_HPP
+#define BOOST_GEOMETRY_EXPERIMENTAL_QUADTREE_PARTITION_HPP
+
+#include <boost/range.hpp>
+#include <boost/geometry/algorithms/detail/partition.hpp>
+
+#include <functional>
+
+namespace boost { namespace geometry
+{
+
+namespace experimental
+{
+
+///! Adapts a lambda of the main visitor to the two-ranges partition algorithm
+template <typename F>
+struct adapt_partition_visitor
+{
+    const F& m_f;
+
+    explicit adapt_partition_visitor(const F& f)
+        : m_f(f)
+    {}
+
+    template<typename T1, typename T2>
+    bool apply(T1 const& i1, T2 const& i2) const
+    {
+        return m_f(i1, i2);
+    }
+};
+
+template <typename F>
+struct adapt_partition_expand_visitor
+{
+    const F& m_f;
+
+    explicit adapt_partition_expand_visitor(const F& f)
+        : m_f(f)
+    {}
+
+    template<typename T1, typename T2>
+    void apply(T1& i1, T2 const& i2) const
+    {
+        m_f(i1, i2);
+    }
+};
+
+template <typename F>
+struct adapt_partition_overlaps_visitor
+{
+    const F& m_f;
+
+    explicit adapt_partition_overlaps_visitor(const F& f)
+        : m_f(f)
+    {}
+
+    template<typename T1, typename T2>
+    bool apply(T1 const& i1, T2 const& i2) const
+    {
+        return m_f(i1, i2);
+    }
+};
+
+template <typename F>
+struct adapt_partition_box_visitor
+{
+    const F& m_f;
+
+    explicit adapt_partition_box_visitor(const F& f)
+        : m_f(f)
+    {}
+
+    // TODO: it would be useful to specify Dimension, and the role the box plays
+    // (normal, exceeding).
+    template<typename Box>
+    void apply(Box const& box, int level) const
+    {
+        m_f(box, level);
+    }
+};
+
+template
+<
+    typename BoxType,
+    typename ForwardRange1,
+    typename ForwardRange2
+>
+bool quadtree_partition(ForwardRange1 const& forward_range1,
+                        ForwardRange2 const& forward_range2,
+                        const std::function
```

Comment:
I'd avoid using `std::function` because it's heavy. Why did you decide to use it over a function object?